File: day4.py
lines = open("day4_input.txt").read().splitlines()
cols = ["".join(line[i] for line in lines) for i in range(len(lines[0]))]
diags = ["".join(c for i, line in enumerate(lines) for j, c in enumerate(line) if i + j == s) for s in range(2*len(lines)-1)] +\
        ["".join(c for i, line in enumerate(lines) for j, c in enumerate(line) if i - j == d) for d in range(-len(lines)+1, len(lines))]
h_words = [line[i:i+4] for line in lines for i in range(len(line)-3)]
v_words = [col[i:i+4] for col in cols for i in range(len(col)-3)]
d_words = [diag[i:i+4] for diag in diags for i in range(len(diag)-3)]
print(sum(len([word for word in group if word in ["XMAS", "SAMX"]]) for group in [h_words, v_words, d_words]))

File: day4.hs
main = do  
    file_content <- readFile "day4_input.txt"
    let lines_ = lines file_content
    let cols = [[line!!i | line <- lines_] | i <- [0..length (lines_!!0) - 1]]
    let diags = [[lines_!!i!!j | i <- [0..length lines_ - 1], j <- [0..length lines_ - 1], i + j == s] | s <- [0..2*(length lines_) - 1]] ++
                [[lines_!!i!!j | i <- [0..length lines_ - 1], j <- [0..length lines_ - 1], i - j == d] | d <- [-(length lines_) + 1..length lines_ - 1]]
    let h_words = [take 4 (drop i line) | line <- lines_, i <- [0..length line - 4]]
    let v_words = [take 4 (drop i col)  | col  <- cols,   i <- [0..length col  - 4]]
    let d_words = [take 4 (drop i diag) | diag <- diags,  i <- [0..length diag - 4]]
    print $ sum [length [word | word <- group, word `elem` ["XMAS", "SAMX"]] | group <- [h_words, v_words, d_words]]